Add tests for directory input CLI feature

turkaturki opened the pull request:
## Description
This PR adds test coverage for the directory input feature implemented in PR #1499, which allows necessist to accept directory paths as CLI input. It creates the test fixtures requested in issue #1511.

## Changes
- Add a test fixture with the following structure:
  - `fixtures/cli_dir_input/target_dir/` - A Rust project containing some test code
  - `fixtures/cli_dir_input/adjacent_dir/` - A separate Rust project with similar test code
- Add a trycmd test case that:
  - Runs necessist with `--root=fixtures/cli_dir_input/target_dir`
  - Verifies that only candidates in the target_dir (not adjacent_dir) are detected

## Testing
The new test passes when run with:
```sh
cargo test -p necessist -- trycmd::necessist_db_absent::cli_dir_input
```

Manual verification confirms that when necessist is run with:
- `--root=fixtures/cli_dir_input/target_dir` - Only finds candidates in the target directory
- `--root=fixtures/cli_dir_input` - Finds candidates in both subdirectories

This validates that the directory input feature correctly handles directory arguments as requested in issue #1511.

## Related Issues
Closes #1511

turkaturki commented:
@smoelius I've implemented the trycmd test for the directory input CLI feature as requested in issue #1511, but I'm running into a CI failure with the `check_toml_files` test. 

The error occurs at line 126 in `trycmd.rs` with:
```
assertion failed: file_stem.starts_with(example)
```

I think this happened because the test validates that the test filename starts with the directory name in the `--root` argument. My test uses:
- Test files: `cli_dir_input.toml` and `cli_dir_input.stdout`
- Arguments: `--root=fixtures/cli_dir_input/target_dir` 

I see two potential approaches to fix this:

1. Rename the test files to `target_dir.toml` and `target_dir.stdout` to match the target directory name
2. Change the test to use `--root=fixtures/cli_dir_input` and modify the expected output to handle both directories

Since this test specifically validates the directory input feature (that necessist can process a directory path), the naming is a bit of a special case compared to other tests.

Which approach would you prefer I take to resolve this CI issue? The current implementation correctly validates that when a directory is specified, only candidates within that directory are processed.

smoelius commented:
Hey, @turkaturki. Thanks a lot for doing this. I'm getting ready to travel, and I might not get to look at this soon. I'll do my best, though.

smoelius commented on necessist/tests/necessist_db_absent/cli_dir_input.toml:
```diff
@@ -0,0 +1,11 @@
+args = ["--no-sqlite", "--dump-candidates", "--framework=rust", "--root=fixtures/cli_dir_input/target_dir"]
```
I would have expected something like this:
```suggestion
args = ["--no-sqlite", "--dump-candidates", "--root=fixtures/cli_dir_input", "src/target_dir"]
```
The `src/target_dir` by itself is what exercises your addition.

So, in other words, `fixtures/cli_dir_input` is the root of a single Cargo project, and `src/target_dir` and `src/adjacent_dir` live inside it.

You might also consider using:
```
args = ["--no-sqlite", "--dump-candidates", "--root=fixtures/cli_dir_input", "src/target_dir", "src/adjacent_dir/some_file.rs"]
```
The idea would be:
- `src/target_dir` exercises the `is_dir` branch in your new code
- `src/adjacent_dir/some_file.rs` exercises the `is_file` branch in your new code

If you do that though, please also have another file in `src/adjacent_dir` (besides `some_file.rs`) whose tests do not get included.

Please tell me if you already know that what I've written won't work.

Also, my "about to travel" comment still applies. :)

turkaturki replied:
hey @smoelius thanks for the suggestions, I'll try to implement all of them, albeit it would take some time, let me know when you get back from travelling, in the mean time I'll try to work on this on my own.

turkaturki commented:
@smoelius Thanks for your suggestions from before! I've implemented all of them as you recommended:

1. Created a single Cargo project structure in `fixtures/cli_dir_input` with:
   - `src/target_dir` and `src/adjacent_dir` subdirectories containing test code
   - Each directory has its own `src/lib.rs` with similar test patterns

2. Updated the test cases to properly exercise both code paths:
   - `cli_dir_input.toml`: Tests passing only the directory path (`fixtures/cli_dir_input/src/target_dir`)
   - `cli_dir_input-mixed.toml`: Tests both directory and file inputs together

3. Ensured the tests verify that:
   - When only a directory is specified, only candidates from that directory are found
   - When both directory and file are specified, candidates from both locations are found

The test structure now matches your suggested approach with a single project containing both directories, which better tests the directory input feature.

Please review when you're back from your travels - no rush. Let me know if any further changes are needed.

turkaturki commented:
@smoelius I don't think the failing [CI / test (macos-latest, third_party_1) (pull_request)](https://github.com/trailofbits/necessist/actions/runs/14505308781/job/40693605857?pr=1517) is due to my code changes, please have a look when you get back.

turkaturki commented:
@smoelius take a look at this when you're free

smoelius commented on necessist/tests/necessist_db_absent/cli_dir_input-mixed.toml:
```diff
@@ -0,0 +1,11 @@
+args = ["--no-sqlite", "--dump-candidates", "--framework=rust", "--root=fixtures/cli_dir_input", "fixtures/cli_dir_input/src/target_dir", "fixtures/cli_dir_input/src/adjacent_dir/src/lib.rs"]
```
```suggestion
args = ["--no-sqlite", "--dump-candidates", "--root=fixtures/cli_dir_input", "fixtures/cli_dir_input/src/target_dir", "fixtures/cli_dir_input/src/adjacent_dir/src/lib.rs"]
```

I don't think `--framework=rust` should be necessary. Does your experience suggest it is necessary?

smoelius commented on fixtures/cli_dir_input/src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+// This is the main lib.rs file for the cli_dir_input test fixture
+// It doesn't need to contain any actual test code as we're testing
+// the directory input feature 
```
```suggestion
// the directory input feature 

mod adjacent_mod;
mod target_mod;
```
Please see my "file structure" comment.

smoelius commented on necessist/tests/necessist_db_absent/cli_dir_input.toml:
```diff
@@ -0,0 +1,11 @@
+args = ["--no-sqlite", "--dump-candidates", "--framework=rust", "--root=fixtures/cli_dir_input", "fixtures/cli_dir_input/src/target_dir"]
```
```suggestion
args = ["--no-sqlite", "--dump-candidates", "--framework=rust", "--root=fixtures/cli_dir_input"]
```

Please see my "two tests" comment.

smoelius reviewed (commented):
I don't think there is much advantage to having a test that passes
```
fixtures/cli_dir_input/src/target_dir
```
and a test that passes
```
fixtures/cli_dir_input/src/target_dir fixtures/cli_dir_input/src/adjacent_dir/src/lib.rs
```
because the latter accomplishes everything the former does. (Apologies, I see now that https://github.com/trailofbits/necessist/pull/1517#discussion_r2043832194 was confusing. Also, this is my "two tests" comment referenced below.)

However, it appears that Necessist has no cli input test currently. So let's please use this PR to add it.

Please change the former test to pass no files or directories. In this way, the former test will act as the control (in the scientific method sense) and the latter test will act as the variable.

---

I think the `lib.rs` files in the `target_dir` and `adjacent_dir` directories are currently unused by Rust. (This is my "file structure" comment referenced below.)

To have them be used, I think the project needs to be structured as follows:
```
src
|
+- lib.rs
|
+- target_dir
|  +- mod.rs
|
+- adjacent_dir
   +- mod.rs
   +- adjacent_file.rs (does not yet exist)
```
Please note that `adjacent_dir/mod.rs` will need to include:
```
mod adjacent_file;
```
Finally, I worry that `target_dir` sounds too much like [Cargo's `target` directory](https://doc.rust-lang.org/cargo/appendix/glossary.html?highlight=target-dir#target).

Could I suggest changing `target` to `some` and `adjacent` to `other` everywhere?

---

Please forgive me, I will likely have to give this PR another look following these changes.

turkaturki commented:
@smoelius thanks for clearing up the file structure for me, I think I successfully implemented all your suggested changes this time.

turkaturki commented:
@smoelius I'll update this branch once #1523 is merged, in the mean time could you please go through my changes and make sure it aligns with your suggestions.

smoelius commented on fixtures/cli_input/src/other_dir/src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+// This is just a placeholder lib.rs file that maintains
```
This file can be removed.

smoelius commented on fixtures/cli_input/src/some_dir/src/lib.rs:
```diff
@@ -0,0 +1,3 @@
+// This is just a placeholder lib.rs file that maintains
```
This file can be removed.

smoelius commented on fixtures/cli_input/Cargo.toml:
```diff
@@ -0,0 +1,8 @@
+edition = "2024"
+
+[workspace]
+
+[dependencies] 
```
```suggestion
```

smoelius commented on fixtures/cli_input/src/other_dir/other_file.rs:
```diff
@@ -0,0 +1,20 @@
+
+#[allow(dead_code)]
+fn other_file_noop() {}
+
+#[test]
+fn unused_other_test() {
+    let value = unused_other_expression();
+    assert!(value);
+}
+
+#[allow(dead_code)]
+fn unused_other_expression() -> bool {
+    true
+} 
```
```suggestion
```
Did you have a specific reason for including these? If not, let's please remove them, just to keep things simple.

smoelius commented on fixtures/cli_input/src/some_dir/mod.rs:
```diff
@@ -0,0 +1,20 @@
+
+#[allow(dead_code)]
+fn some_noop() {}
+
+#[test]
+fn unused_test() {
+    let value = unused_expression();
+    assert!(value);
+}
+
+#[allow(dead_code)]
+fn unused_expression() -> bool {
+    true
+} 
```
```suggestion
```
Same comment as above: if you didn't have a specific reason for including these, let's please remove them.

smoelius commented on fixtures/cli_input/src/other_dir/mod.rs:
```diff
@@ -0,0 +1,12 @@
+fn other_noop() {}
+
+// Add a reference to the adjacent_file module as requested by the reviewer
+mod other_file; 
```
```suggestion
mod other_file; 
mod some_file; 
```
Sorry, I'm being pedantic now. But for symmetry, let's please move the tests from this file into `some_file.rs`.

smoelius commented on necessist/tests/necessist_db_absent/cli_input-mixed.toml:
```diff
@@ -0,0 +1,11 @@
+args = ["--no-sqlite", "--dump-candidates", "--root=fixtures/cli_input", "fixtures/cli_input/src/some_dir", "fixtures/cli_input/src/other_dir/other_file.rs"]
```
```suggestion
args = ["--no-sqlite", "--dump-candidates", "--root=fixtures/cli_input", "fixtures/cli_input/src/some_dir", "fixtures/cli_input/src/other_dir/some_file.rs"]
```
Nit (so that the selected dir and file are the "some" ones).

smoelius commented on necessist/tests/necessist_db_absent/cli_input-mixed.stdout:
```diff
@@ -0,0 +1,2 @@
+fixtures/cli_input/src/other_dir/other_file.rs:4:5-4:12: `n += 1;`
```
Nit. Please change the `-` to `_` in these filenames.

smoelius reviewed (commented):
@turkaturki Thanks for your patience. I think we're just about there.

turkaturki commented:
Thanks for the detailed suggestion @smoelius. I made the changes you suggested, have a look to see if I missed anything. I just noticed that you have a lot of PRs in dylint, so I don't want to take much of your time.

smoelius commented:
Just one more nit. Several of the files are missing newlines at the end. You can see that in GH's "Files changed" view: https://github.com/trailofbits/necessist/pull/1517/files

Could you please add the newlines? Then I will merge this.

smoelius commented:
@turkaturki Is this good to go?

turkaturki commented:
@smoelius yeah I think so.

smoelius commented:
This is great! Thanks A LOT!
